# Dreem Project - Deep Learning - TF & Keras - Basile NOUVELLET

## Check if GPU is used

In [1]:
from tensorflow.python.client import device_lib
print(device_lib.list_local_devices())

[name: "/device:CPU:0"
device_type: "CPU"
memory_limit: 268435456
locality {
}
incarnation: 8909690045639871536
, name: "/device:XLA_GPU:0"
device_type: "XLA_GPU"
memory_limit: 17179869184
locality {
}
incarnation: 9604628808204645324
physical_device_desc: "device: XLA_GPU device"
, name: "/device:XLA_CPU:0"
device_type: "XLA_CPU"
memory_limit: 17179869184
locality {
}
incarnation: 16661797590608488637
physical_device_desc: "device: XLA_CPU device"
, name: "/device:GPU:0"
device_type: "GPU"
memory_limit: 11330115994
locality {
  bus_id: 1
  links {
  }
}
incarnation: 6979578593716129426
physical_device_desc: "device: 0, name: Tesla K80, pci bus id: 0000:00:1e.0, compute capability: 3.7"
]


## Data

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

In [3]:
df_train_final = pd.read_hdf('df_train_final.h5', 'df_train_final')

var_to_pred = 'SO'
X_train, X_val, y_train, y_val = train_test_split(df_train_final.loc[:, df_train_final.columns != var_to_pred],
                                                  df_train_final[var_to_pred],
                                                  test_size=0.10,
                                                  random_state=0,
                                                  stratify=df_train_final[var_to_pred])
y_train = pd.DataFrame(y_train)
y_val = pd.DataFrame(y_val)

In [4]:
df_train_final.head()

num_pso  mean_amp_pso  mean_dur_pso     amp_cso  dur_cso  time_since_sleep  \
0    237.0    152.658761    341.523207  128.017491    429.0           11379.0   
1    176.0    146.883435    338.039773  119.130849    196.0            6721.0   
2    456.0    152.376541    335.629386  164.292580    417.0           26832.0   
3     21.0    139.720772    336.285714  159.237082    407.0            1289.0   
4     72.0    140.649432    349.875000  130.184278    297.0            2262.0   

   time_in_ds  time_in_ls  time_in_rs  time_in_ws  ...  wawelets_86  \
0      2730.0      3780.0         0.0       480.0  ...          457   
1      2580.0      2100.0         0.0       480.0  ...          453   
2      3240.0      7440.0      2130.0       750.0  ...          462   
3         0.0        60.0         0.0       450.0  ...          460   
4       630.0       960.0         0.0       450.0  ...          452   

   wawelets_87  wawelets_88  wawelets_89  wawelets_90  wawelets_91  \
0    -1.508171    -0.600457     0.605929     1.524894     0.011170   
1    -1.302524    -0.578372     0.553759     1.361592    -0.032801   
2    -1.774095    -0.736667     0.757092     1.778728     0.016140   
3    -1.514035    -0.667986     0.661924     1.535148    -0.005341   
4    -1.425063    -0.590922     0.636247     1.404894     0.001580   

   wawelets_92  wawelets_93  wawelets_94  wawelets_95  
0    -0.000105     1.006676     1.013396     0.740114  
1     0.001091     0.845156     0.714289     0.661250  
2    -0.001139     1.099851     1.209672     0.872444  
3     0.000916     0.961756     0.924975     0.773179  
4    -0.002912     0.872183     0.760704     0.701210  

[5 rows x 181 columns]

In [5]:
features = list(df_train_final.columns)

equations = {
     'base_model': {'predictors': features[0:13]},
     'entropy_fft': {'predictors': features[0:81]},
     'dummy_variables_only': {'predictors': features[0:11] + features[81:83]},
     'dummy_variables': {'predictors': features[0:83]},
     'fft_wawelets' : {'predictors': features[0:81] + features[85:181]}
}

features

['num_pso',
 'mean_amp_pso',
 'mean_dur_pso',
 'amp_cso',
 'dur_cso',
 'time_since_sleep',
 'time_in_ds',
 'time_in_ls',
 'time_in_rs',
 'time_in_ws',
 'mean',
 'max',
 'min',
 'perm_entropy',
 'svd_entropy',
 'higuchi_fd',
 'detrended_fluctuation',
 'fft_0',
 'fft_1',
 'fft_2',
 'fft_3',
 'fft_4',
 'fft_5',
 'fft_6',
 'fft_7',
 'fft_8',
 'fft_9',
 'fft_10',
 'fft_11',
 'fft_12',
 'fft_13',
 'fft_14',
 'fft_15',
 'fft_16',
 'fft_17',
 'fft_18',
 'fft_19',
 'fft_20',
 'fft_21',
 'fft_22',
 'fft_23',
 'fft_24',
 'fft_25',
 'fft_26',
 'fft_27',
 'fft_28',
 'fft_29',
 'fft_30',
 'fft_31',
 'fft_32',
 'fft_33',
 'fft_34',
 'fft_35',
 'fft_36',
 'fft_37',
 'fft_38',
 'fft_39',
 'fft_40',
 'fft_41',
 'fft_42',
 'fft_43',
 'fft_44',
 'fft_45',
 'fft_46',
 'fft_47',
 'fft_48',
 'fft_49',
 'fft_50',
 'fft_51',
 'fft_52',
 'fft_53',
 'fft_54',
 'fft_55',
 'fft_56',
 'fft_57',
 'fft_58',
 'fft_59',
 'fft_60',
 'fft_61',
 'fft_62',
 'fft_63',
 'curr_sleep_stage_2.0',
 'curr_sleep_stage_3.0',
 'slow

## Special data treatment for Keras

In [6]:
from keras.utils import to_categorical

# Convert labels to categorical one-hot encoding
y_train = to_categorical(y_train.values, num_classes=3)
y_val = to_categorical(y_val.values, num_classes=3)

Using TensorFlow backend.


## Same architecture on all models

In [7]:
from keras.models import Sequential
from keras.layers import Dense, Dropout

In [8]:
epochs = 5
batch_size = 128

results = pd.DataFrame(columns=['model_name', 'loss', 'accuracy'])

for model_name in equations:
    predictors = equations[model_name]['predictors']

    x_train = X_train[predictors].values
    x_val = X_val[predictors].values

    model = Sequential([
        Dense(units=10, activation='relu', input_dim=x_train.shape[1]),
        Dropout(0.2),
        Dense(units=10, activation='relu'),
        Dropout(0.2),
        Dense(units=10, activation='relu'),
        Dropout(0.2),
        Dense(units=3, activation='softmax'),
    ])
    
    model.compile(loss='categorical_crossentropy',
                  optimizer='sgd',
                  metrics=['accuracy'])

    model.fit(x_train, y_train, epochs=epochs, batch_size=batch_size)

    loss_and_metrics = model.evaluate(x_val, y_val)

    print("Result for %s: %s\n\n" % (model_name, loss_and_metrics))

    results = results.append({
        'model_name': model_name,
        'loss': loss_and_metrics[0],
        'accuracy': loss_and_metrics[1],
    }, ignore_index=True)





Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.


Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Epoch 1/5
235470/235470 [==============================] - 61s 257us/step - loss: 11.2093 - acc: 0.2964
Epoch 2/5
235470/235470 [==============================] - 7s 31us/step - loss: 9.8606 - acc: 0.3206
Epoch 3/5
235470/235470 [==============================] - 7s 29us/step - loss: 8.9551 - acc: 0.3308
Epoch 4/5
235470/235470 [==============================] - 7s 29us/step - loss: 8.9359 - acc: 0.3325
Epoch 5/5
26164/26164 [==============================] - 1s 39us/step
Result for base_model: [11.2639201764502, 0.3009478673076907]


Epoch 1/5
235470/235470 [==============================] - 7s 31us/step - loss: 9.1990 - acc: 0.3062
Epoch 2/5
235470/235470 [==============================] - 7s 29us/step - loss: 8.1490 - acc: 0.3238
Epoch 3/5
235470/235470 [=================

In [9]:
results

model_name       loss  accuracy
0            base_model  11.263920  0.300948
1           entropy_fft  11.868645  0.263645
2  dummy_variables_only   1.074749  0.435866
3       dummy_variables   1.074756  0.435866
4          fft_wawelets  11.868645  0.263645

In [10]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_17 (Dense)             (None, 10)                1780      
_________________________________________________________________
dropout_13 (Dropout)         (None, 10)                0         
_________________________________________________________________
dense_18 (Dense)             (None, 10)                110       
_________________________________________________________________
dropout_14 (Dropout)         (None, 10)                0         
_________________________________________________________________
dense_19 (Dense)             (None, 10)                110       
_________________________________________________________________
dropout_15 (Dropout)         (None, 10)                0         
_________________________________________________________________
dense_20 (Dense)             (None, 3)                 33        
Total para

## Separate curr_sleep 2 & 3

In [11]:
def get_train_test_split(df):
    return train_test_split(df.loc[:, df.columns != var_to_pred],
                            df[var_to_pred],
                            test_size=0.10,
                            random_state=0,
                            stratify=df[var_to_pred])

In [12]:
df_train_final_2 = df_train_final[ df_train_final["curr_sleep_stage_2.0"] == 1 ]
df_train_final_3 = df_train_final[ df_train_final["curr_sleep_stage_3.0"] == 1 ]

var_to_pred = 'SO'

# Curr sleep 2
X_train_2, X_val_2, y_train_2, y_val_2 = get_train_test_split(df_train_final_2)

y_train_2 = to_categorical(pd.DataFrame(y_train_2).values, num_classes=3)
y_val_2 = to_categorical(pd.DataFrame(y_val_2).values, num_classes=3)

# Curr sleep 3
X_train_3, X_val_3, y_train_3, y_val_3 = get_train_test_split(df_train_final_3)

y_train_3 = to_categorical(pd.DataFrame(y_train_3).values, num_classes=3)
y_val_3 = to_categorical(pd.DataFrame(y_val_3).values, num_classes=3)

In [13]:
epochs = 60
batch_size = 512

model_name = "fft_wawelets"

predictors = equations[model_name]['predictors']
print("Model is '%s' (%d predictors)\n" % (model_name, len(predictors)))

x_train_2 = X_train_2[predictors].values
x_val_2 = X_val_2[predictors].values

x_train_3 = X_train_3[predictors].values
x_val_3 = X_val_3[predictors].values

architecture = [
    Dense(units=150, activation='relu', input_dim=x_train_2.shape[1]),
    Dropout(0.2),
    Dense(units=100, activation='relu'),
    Dropout(0.3),
    Dense(units=60, activation='relu'),
    Dropout(0.4),
    Dense(units=30, activation='relu'),
    Dropout(0.3),
    Dense(units=3, activation='softmax'),
]

model_2 = Sequential(architecture)
model_3 = Sequential(architecture)

model_2.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model_3.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

model_2.fit(x_train_2, y_train_2, epochs=epochs, batch_size=batch_size)
model_3.fit(x_train_3, y_train_3, epochs=epochs, batch_size=batch_size)

loss_and_metrics_2 = model_2.evaluate(x_val_2, y_val_2)
loss_and_metrics_3 = model_3.evaluate(x_val_3, y_val_3)

print("\nResult for %s (2): %s" % (model_name, loss_and_metrics_2))
print("Result for %s (3): %s" % (model_name, loss_and_metrics_3))

Model is 'fft_wawelets' (177 predictors)

Epoch 1/60
118229/118229 [==============================] - 3s 28us/step - loss: 6.9936 - acc: 0.5660
Epoch 2/60
118229/118229 [==============================] - 2s 20us/step - loss: 6.8364 - acc: 0.5758
Epoch 3/60
118229/118229 [==============================] - 2s 20us/step - loss: 6.7924 - acc: 0.5786
Epoch 4/60
118229/118229 [==============================] - 2s 19us/step - loss: 6.7846 - acc: 0.5791
Epoch 5/60
118229/118229 [==============================] - 2s 19us/step - loss: 6.7806 - acc: 0.5793
Epoch 6/60
118229/118229 [==============================] - 2s 19us/step - loss: 6.7801 - acc: 0.5794
Epoch 7/60
118229/118229 [==============================] - 2s 20us/step - loss: 6.7810 - acc: 0.5793
Epoch 8/60
118229/118229 [==============================] - 2s 20us/step - loss: 6.7812 - acc: 0.5793
Epoch 9/60
118229/118229 [==============================] - 2s 19us/step - loss: 6.7816 - acc: 0.5793
Epoch 10/60
118229/118229 [=============

##### Epochs=5, batch_size=128

2: 0.579

3: 0.368


##### Epochs=60, batch_size=512

2: 0.233

3: 0.368

## Original dataset (NO NEED TO RE-RUN)

In [36]:
def label_columns(df):
    l_columns = [
        'num_pso', 
        'mean_amp_pso',
        'mean_dur_pso',
        'amp_cso',
        'dur_cso',
        'curr_sleep_stage',
        'time_since_sleep',
        'time_in_ds',
        'time_in_ls',
        'time_in_rs',
        'time_in_ws',
    ]

    for i in range(12, 1261+1):
        l_columns.append('eeg_signal_%s' % (i - 12 + 1))

    df.columns = l_columns

    return df

In [47]:
import h5py

y_train = pd.read_csv("original_data/y_train.csv").values[:, 1].squeeze()
X_train = h5py.File("original_data/X_train.h5", "r")

df_train = pd.DataFrame(data=X_train["features"][:])
label_columns(df_train).head()

num_pso  mean_amp_pso  mean_dur_pso     amp_cso  dur_cso  curr_sleep_stage  \
0    237.0    152.658761    341.523207  128.017491    429.0               3.0   
1    176.0    146.883435    338.039773  119.130849    196.0               2.0   
2    456.0    152.376541    335.629386  164.292580    417.0               2.0   
3     21.0    139.720772    336.285714  159.237082    407.0               3.0   
4     72.0    140.649432    349.875000  130.184278    297.0               3.0   

   time_since_sleep  time_in_ds  time_in_ls  time_in_rs  ...  eeg_signal_1241  \
0           11379.0      2730.0      3780.0         0.0  ...        22.707487   
1            6721.0      2580.0      2100.0         0.0  ...         2.839801   
2           26832.0      3240.0      7440.0      2130.0  ...        27.093293   
3            1289.0         0.0        60.0         0.0  ...        28.191571   
4            2262.0       630.0       960.0         0.0  ...        -7.942308   

   eeg_signal_1242  eeg_signal_1243  eeg_signal_1244  eeg_signal_1245  \
0        15.042639         9.706864         7.305480         5.344436   
1         3.458973         4.441102         3.975107         1.000240   
2        22.821611        14.196937         5.708701        -0.753271   
3        25.032295        17.296456         6.335396        -4.008689   
4       -10.407617        -3.418883         9.222596        21.236168   

   eeg_signal_1246  eeg_signal_1247  eeg_signal_1248  eeg_signal_1249  \
0         2.674903        -0.055816        -1.212385        -2.461937   
1         2.870631         7.071897         7.848365         4.033517   
2        -5.627993        -9.804085       -12.863908       -11.951175   
3        -8.589818        -5.876062         0.166707         6.054539   
4        28.245889        27.024864        17.794644         7.602379   

   eeg_signal_1250  
0        -4.930397  
1        -2.110046  
2        -5.531799  
3        12.086351  
4        -4.548318  

[5 rows x 1261 columns]

In [48]:
df_train.to_hdf('df_train.h5', key='df_train')

## Model with original dataset

In [ ]:
df_train = pd.read_hdf('df_train.h5', 'df_train')